File: app/src/main/java/jp/juggler/subwaytooter/api/TootTaskRunner.java
package jp.juggler.subwaytooter.api;

import android.app.Activity;
import android.app.ProgressDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.os.AsyncTask;
import android.os.Handler;
import android.os.SystemClock;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.text.TextUtils;

import java.lang.ref.WeakReference;
import java.text.NumberFormat;

import jp.juggler.subwaytooter.App1;
import jp.juggler.subwaytooter.table.SavedAccount;

/*
	非同期タスク(TootTask)を実行します。
	- 内部でAsyncTaskを使います。Android Lintの警告を抑制します
	- ProgressDialogを表示します。抑制することも可能です。
	- TootApiClientの初期化を行います
	- TootApiClientからの進捗イベントをProgressDialogに伝達します。
*/

public class TootTaskRunner extends TootApiClient.Callback {
	
	@SuppressWarnings("WeakerAccess") public static final int PROGRESS_NONE = - 1;
	@SuppressWarnings("WeakerAccess") public static final int PROGRESS_SPINNER = ProgressDialog.STYLE_SPINNER;
	@SuppressWarnings("WeakerAccess") public static final int PROGRESS_HORIZONTAL = ProgressDialog.STYLE_HORIZONTAL;
	
	private static int getDefaultProgressStyle( boolean bShowProgress ){
		return bShowProgress ? PROGRESS_SPINNER : PROGRESS_NONE;
	}
	
	private static class ProgressInfo {
		
		// HORIZONTALスタイルの場合、初期メッセージがないと後からメッセージを指定しても表示されない
		@NonNull String message = " ";
		
		boolean isIndeterminate = true;
		int value = 0;
		int max = 1;
	}
	
	@NonNull protected final Handler handler;
	@NonNull protected final TootApiClient client;
	@NonNull private final MyTask task;
	@NonNull protected final ProgressInfo info = new ProgressInfo();
	@Nullable private ProgressDialog progress;
	@Nullable private String progress_prefix;
	
	@NonNull private final WeakReference<Context> refContext;
	private final int progress_style;
	
	private TootTask callback;
	private long last_message_shown;
	
	private static final NumberFormat percent_format;
	
	static{
		percent_format = NumberFormat.getPercentInstance();
		percent_format.setMaximumFractionDigits( 0 );
	}
	
	public TootTaskRunner( @NonNull Context context ){
		this( context,PROGRESS_NONE);
	}

	public TootTaskRunner( @NonNull Context context, boolean bShowProgress ){
		this( context, getDefaultProgressStyle( bShowProgress ) );
	}

	public TootTaskRunner( @NonNull Context context, int progress_style ){
		this.refContext = new WeakReference<>( context );
		this.progress_style = progress_style;
		this.handler = new Handler();
		this.client = new TootApiClient( context, this );
		this.task = new MyTask(this);
	}

	public void run( @NonNull TootTask callback){
		openProgress();
		
		this.callback = callback;
		
		task.executeOnExecutor( App1.task_executor );
	}

	public void run( @NonNull SavedAccount access_info,@NonNull TootTask callback){
		client.setAccount( access_info );
		run(callback);
	}

	public void run( @NonNull String instance,@NonNull TootTask callback){
		client.setInstance( instance );
		run(callback);
	}
	
	public TootTaskRunner progressPrefix( String s ){
		this.progress_prefix = s;
		return this;
	}
	
	//////////////////////////////////////////////////////
	// has AsyncTask

	protected static class MyTask extends AsyncTask< Void, Void, TootApiResult >{
		
		private TootTaskRunner runner;

		MyTask(TootTaskRunner runner){
			this.runner = runner;
		}

		@Override protected TootApiResult doInBackground( Void... voids ){
			return runner.callback.background( runner.client );
		}
		
		@Override protected final void onCancelled( TootApiResult result ){
			onPostExecute( result );
		}
		
		@Override protected final void onPostExecute( TootApiResult result ){
			runner.dismissProgress();
			runner.callback.handleResult( result );
		}
	}
	
	//////////////////////////////////////////////////////
	// implements TootApiClient.Callback
	
	@Override public boolean isApiCancelled(){
		return task.isCancelled();
	}
	
	@Override public void publishApiProgress( @NonNull final String s ){
		synchronized( this ){
			info.message = s;
			info.isIndeterminate = true;
		}
		delayProgressMessage();
	}
	
	@Override public void publishApiProgressRatio( final int value, final int max ){
		synchronized( this ){
			info.isIndeterminate = false;
			info.value = value;
			info.max = max;
		}
		delayProgressMessage();
	}
	
	//////////////////////////////////////////////////////
	// ProgressDialog
	
	private void openProgress(){
		// open progress
		if( progress_style != PROGRESS_NONE ){
			Context context = refContext.get();
			if( context != null && context instanceof Activity ){
				//noinspection deprecation
				progress = new ProgressDialog( context );
				progress.setCancelable( true );
				progress.setOnCancelListener( new DialogInterface.OnCancelListener() {
					@Override public void onCancel( DialogInterface dialog ){
						task.cancel( true );
					}
				} );
				progress.setProgressStyle( progress_style );
				showProgressMessage();
				progress.show();
			}
		}
	}
	
	// ダイアログを閉じる
	private void dismissProgress(){
		if( progress != null ){
			try{
				progress.dismiss();
			}catch( Throwable ignored ){
				// java.lang.IllegalArgumentException:
				// at android.view.WindowManagerGlobal.findViewLocked(WindowManagerGlobal.java:396)
				// at android.view.WindowManagerGlobal.removeView(WindowManagerGlobal.java:322)
				// at android.view.WindowManagerImpl.removeViewImmediate(WindowManagerImpl.java:116)
				// at android.app.Dialog.dismissDialog(Dialog.java:341)
				// at android.app.Dialog.dismiss(Dialog.java:324)
			}finally{
				progress = null;
			}
		}
	}
	
	// ダイアログのメッセージを更新する
	// 初期化時とメッセージ更新時に呼ばれる
	private void showProgressMessage(){
		if( progress == null ) return;
		
		if( ! TextUtils.isEmpty( progress_prefix ) ){
			progress.setMessage( progress_prefix + ( TextUtils.isEmpty( info.message.trim() ) ? "" : "\n" + info.message ) );
		}else{
			progress.setMessage( info.message );
		}
		
		progress.setIndeterminate( info.isIndeterminate );
		if( info.isIndeterminate ){
			progress.setProgressNumberFormat( null );
			progress.setProgressPercentFormat( null );
		}else{
			progress.setProgress( info.value );
			progress.setMax( info.max );
			progress.setProgressNumberFormat( "%1$,d / %2$,d" );
			progress.setProgressPercentFormat( percent_format );
		}
		
		last_message_shown = SystemClock.elapsedRealtime();
	}

	// 少し後にダイアログのメッセージを更新する
	// あまり頻繁に更新せず、しかし繰り返し呼ばれ続けても時々は更新したい
	// どのスレッドから呼ばれるか分からない
	private void delayProgressMessage(){
		long wait = 100L + last_message_shown - SystemClock.elapsedRealtime();
		wait = wait < 0L ? 0L : wait > 100L ? 100L : wait;

		synchronized( this ){
			handler.removeCallbacks( proc_progress_message );
			handler.postDelayed( proc_progress_message, wait );
		}
	}
	
	private final Runnable proc_progress_message = new Runnable() {
		@Override public void run(){
			synchronized( this ){
				if( progress != null && progress.isShowing() ){
					showProgressMessage();
				}
			}
		}
	};
	
}

File: app/src/main/java/jp/juggler/subwaytooter/api/TootTaskRunner.kt
package jp.juggler.subwaytooter.api

import android.app.Activity
import android.content.Context
import android.os.Handler
import android.os.SystemClock
import jp.juggler.subwaytooter.App1
import jp.juggler.subwaytooter.api.entity.Host
import jp.juggler.subwaytooter.dialog.ProgressDialogEx
import jp.juggler.subwaytooter.table.SavedAccount
import jp.juggler.util.LogCategory
import jp.juggler.util.clip
import jp.juggler.util.dismissSafe
import jp.juggler.util.withCaption
import kotlinx.coroutines.*
import java.lang.Runnable
import java.lang.ref.WeakReference
import java.text.NumberFormat
import java.util.concurrent.atomic.AtomicBoolean

/*
	APIクライアントを必要とする非同期タスク(TootTask)を実行します。
	- ProgressDialogを表示します。抑制することも可能です。
	- TootApiClientの初期化を行います
	- TootApiClientからの進捗イベントをProgressDialogに伝達します。
*/

class TootTaskRunner(
	context: Context,
	private val progress_style: Int = PROGRESS_SPINNER,
	private val progressSetupCallback: (progress: ProgressDialogEx) -> Unit = { _ -> }
) : TootApiCallback {

    companion object {
        private val log = LogCategory("TootTaskRunner")

        const val PROGRESS_NONE = -1
        const val PROGRESS_SPINNER = ProgressDialogEx.STYLE_SPINNER
        const val PROGRESS_HORIZONTAL = ProgressDialogEx.STYLE_HORIZONTAL

        private val percent_format: NumberFormat by lazy {
            val v = NumberFormat.getPercentInstance()
            v.maximumFractionDigits = 0
            v
        }
    }

    private class ProgressInfo {

        // HORIZONTALスタイルの場合、初期メッセージがないと後からメッセージを指定しても表示されない
        var message = " "
        var isIndeterminate = true
        var value = 0
        var max = 1
    }

    private val handler: Handler
    private val client: TootApiClient
    private val info = ProgressInfo()
    private var progress: ProgressDialogEx? = null
    private var progress_prefix: String? = null
    private var task: Deferred<TootApiResult?>? = null

    private val refContext: WeakReference<Context>

    private var last_message_shown: Long = 0

    private val proc_progress_message = Runnable {
		if (progress?.isShowing == true) showProgressMessage()
	}

    init {
        this.refContext = WeakReference(context)
        this.handler = App1.getAppState(context, "TootTaskRunner.ctor").handler
        this.client = TootApiClient(context, callback = this)
    }


    val isActive: Boolean
        get() = task?.isActive ?: true // nullはまだ開始してないのでアクティブということにする

    fun run(callback: TootTask) = this.also {
        GlobalScope.launch(Dispatchers.Main) {
            callback.handleResult(
				try {
					openProgress()
					async(Dispatchers.IO) {
						callback.background(client)
					}.also {
						this@TootTaskRunner.task = it
					}.await()
				} catch (ex: CancellationException) {
					null
				} catch (ex: Throwable) {
					TootApiResult(ex.withCaption("error"))
				} finally {
					dismissProgress()
				}
			)
        }
    }

    fun run(access_info: SavedAccount, callback: TootTask): TootTaskRunner {
        client.account = access_info
        return run(callback)
    }

    fun run(instance: Host, callback: TootTask): TootTaskRunner {
        client.apiHost = instance
        return run(callback)

    }

    fun progressPrefix(s: String): TootTaskRunner {
        this.progress_prefix = s
        return this
    }

    //////////////////////////////////////////////////////
    // implements TootApiClient.Callback

    override val isApiCancelled: Boolean
        get() = task?.isActive == false

    override suspend fun publishApiProgress(s: String) {
        synchronized(this) {
            info.message = s
            info.isIndeterminate = true
        }
        delayProgressMessage()
    }

    override suspend fun publishApiProgressRatio(value: Int, max: Int) {
        synchronized(this) {
            info.isIndeterminate = false
            info.value = value
            info.max = max
        }
        delayProgressMessage()
    }

    //////////////////////////////////////////////////////
    // ProgressDialog

    private fun openProgress() {
        // open progress
        if (progress_style != PROGRESS_NONE) {
            val context = refContext.get()
            if (context != null && context is Activity) {
                val progress = ProgressDialogEx(context)
                this.progress = progress
                progress.setCancelable(true)
                progress.setOnCancelListener { task?.cancel() }
                @Suppress("DEPRECATION")
                progress.setProgressStyle(progress_style)
                progressSetupCallback(progress)
                showProgressMessage()
                progress.show()
            }
        }
    }

    // ダイアログを閉じる
    private fun dismissProgress() {
        progress?.dismissSafe()
        progress = null
    }

    // ダイアログのメッセージを更新する
    // 初期化時とメッセージ更新時に呼ばれる
    @Suppress("DEPRECATION")
    private fun showProgressMessage() {
        val progress = this.progress ?: return

		synchronized(this){
			val message = info.message.trim { it <= ' ' }
			val progress_prefix = this.progress_prefix
			progress.setMessageEx(
				when {
					progress_prefix?.isNotEmpty() != true -> message
					message.isEmpty() -> progress_prefix
					else -> "$progress_prefix\n$message"
				}
			)

			progress.isIndeterminateEx = info.isIndeterminate
			if (info.isIndeterminate) {
				progress.setProgressNumberFormat(null)
				progress.setProgressPercentFormat(null)
			} else {
				progress.progress = info.value
				progress.max = info.max
				progress.setProgressNumberFormat("%1$,d / %2$,d")
				progress.setProgressPercentFormat(percent_format)
			}

			last_message_shown = SystemClock.elapsedRealtime()
		}
    }

    // 少し後にダイアログのメッセージを更新する
    // あまり頻繁に更新せず、しかし繰り返し呼ばれ続けても時々は更新したい
    // どのスレッドから呼ばれるか分からない
    private fun delayProgressMessage() {
        var wait = 100L + last_message_shown - SystemClock.elapsedRealtime()
        wait = wait.clip(0L, 100L)

        synchronized(this) {
            handler.removeCallbacks(proc_progress_message)
            handler.postDelayed(proc_progress_message, wait)
        }
    }
}
